copycat/runtime/src/main/java/org/apache/kafka/copycat/runtime/distributed/KafkaConfigStorage.java

```diff
@@ -0,0 +1,438 @@
+            }
+
+            // Generate new root config
+            newRootConfig.putAll(rootConfig);
+            newRootConfig.putAll(updatedRootConfigEntries);
+
+            // Second half of validation: if we find some missing task configurations even with a) outstanding deferred
+            // updates and b) these updates, then we shouldn't generate an updated root config
+            boolean inconsistent = false;
+            for (Map.Entry<String, Integer> newRootEntry : newRootConfig.entrySet()) {
+                String connName = newRootEntry.getKey();
+                for (int taskIndex = 0; taskIndex < newRootEntry.getValue(); taskIndex++) {
+                    ConnectorTaskId taskId = new ConnectorTaskId(connName, taskIndex);
+                    // Note that we *do not* include deferred updates here. This is important because a half completed
+                    // write of updated task configs could leave some left over data in the deferred tasks, which would
+                    // then make this appear to have a complete set of task configs. In order to ensure correctness,
+                    // this method call or the existing task configs need to ensure everything that is inconsistent is
+                    // addressed in one batch.
+                    Map<String, String> taskConfig = configs.containsKey(taskId) ? configs.get(taskId) : taskConfigs.get(taskId);
+                    if (taskConfig == null) {
+                        log.error("Found inconsistent/incomplete assignment, missing config for task " + taskId);
+                        inconsistent = true;
+                    }
+                }
+            }
+            if (inconsistent)
+                throw new CopycatException("Error writing task configurations: found some connectors with incomplete task assignments");
+        }
+
+        // Start sending all the individual updates
+        for (Map.Entry<ConnectorTaskId, Map<String, String>> taskConfigEntry : configs.entrySet()) {
+            Struct copycatConfig = new Struct(TASK_CONFIGURATION_V0);
+            copycatConfig.put("properties", taskConfigEntry.getValue());
+            byte[] serializedConfig = converter.fromCopycatData(topic, TASK_CONFIGURATION_V0, copycatConfig);
+            configLog.send(TASK_KEY(taskConfigEntry.getKey()), serializedConfig);
+        }
+
+        // Finally, send the root update, then wait until we read to the end of the log
+        try {
+            // Read to end to ensure all the task configs have been written
+            configLog.readToEnd().get(READ_TO_END_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+
+            Struct copycatConfig = new Struct(ROOT_V0);
+            copycatConfig.put("connectors", newRootConfig);
+            byte[] serializedConfig = converter.fromCopycatData(topic, ROOT_V0, copycatConfig);
+            configLog.send(ROOT_KEY, serializedConfig);
+
+            configLog.readToEnd().get(READ_TO_END_TIMEOUT_MS, TimeUnit.MILLISECONDS);
```

Comment:
send-and-read-to-end seems to be a repeating pattern. Perhaps it should be part of the KafkaBasedLog API.